### PR TITLE
Specify application Initialization arguments in JSON

### PR DESCRIPTION
**Pull request by christos-h**
# Motivation

Passing initialisation arguments to the `linera` binary requires hex-encoding and bcs-encoding them which is not a great UX.

# Solution

Introduce `--json-args` and `--file-args` which allow you to provide initialization arguments to a linera app encoded in JSON.

### CODE REVIEW

File: linera-sdk/src/lib.rs [hunk 1]
```diff
@@ -79,9 +80,11 @@ pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
 #[async_trait]
 pub trait Contract: Sized {
     /// Message reports for application execution errors.
-    type Error: Error;
+    type Error: Error + From<serde_json::Error>;
     /// The desired storage backend to use to store the application's state.
     type Storage;
+    /// The initialisation arguments for the contract.
+    type InitArg: DeserializeOwned + Send;
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:87]: Thanks for the changes but this one belongs to another PR. We just want to quickly improve the command line of the `linera` tool here

File: examples/fungible/src/lib.rs [hunk 2]
```diff
@@ -53,14 +53,75 @@ pub enum SessionCall {
...
+                formatter.write_str("a string representing an AccountOwner")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let parts: Vec<&str> = value.splitn(2, ':').collect();
+                if parts.len() != 2 {
+                    return Err(serde::de::Error::invalid_length(parts.len(), &self));
```
Comment (afck) [inline examples/fungible/src/lib.rs:102]: Maybe `Error::custom` would be more appropriate. The actual error here is just that the string didn't contain a colon.

File: examples/fungible/src/lib.rs [hunk 3]
```diff
@@ -53,14 +53,75 @@ pub enum SessionCall {
...
+                }
+
+                match parts[0] {
+                    "User" => {
+                        let owner = Owner::from_str(parts[1]).unwrap();
+                        Ok(AccountOwner::User(owner))
+                    }
+                    "Application" => {
+                        let app_id = ApplicationId::from_str(parts[1]).unwrap();
```
Comment (afck) [inline examples/fungible/src/lib.rs:111]: Those `unwrap`s will probably also become `Error::custom`?

File: examples/reentrant-counter/src/contract.rs [hunk 4]
```diff
@@ -104,4 +104,8 @@ pub enum Error {
     /// Invalid serialized increment value.
     #[error("Invalid serialized increment value")]
     InvalidIncrement(#[from] bcs::Error),
+
+    /// Invalid serialized increment value.
```
Comment (afck) [inline examples/reentrant-counter/src/contract.rs:108]:
```suggestion
    /// Invalid serialized initialization value.
```

File: linera-service/src/linera.rs [hunk 5]
```diff
@@ -762,6 +761,40 @@ enum ProjectCommand {
...
+            (Some(file), None) => {
+                let file_contents = fs::read_to_string(file)?;
+                let as_value: Value = serde_json::from_str(&file_contents)?;
+                Ok(serde_json::to_vec(&as_value)?)
+            }
+            (None, Some(json)) => {
+                let as_value: Value = serde_json::from_str(&json)?;
+                Ok(serde_json::to_vec(&as_value)?)
+            }
+            (None, None) => bail!("json arguments must be provided"),
```
Comment (afck) [inline linera-service/src/linera.rs:793]:
```suggestion
            (None, None) => bail!("JSON arguments must be provided"),
```

Alternatively, we _could_ make them optional?

Comment (christos-h) [reply linera-service/src/linera.rs:793]: Indeed, I don't think we have consensus on this yet so I'm leaving it for a future discussion 😅

File: examples/fungible/src/lib.rs (hunk 3 above)
Comment (christos-h) [reply examples/fungible/src/lib.rs]: Yup :)

**Review by afck — APPROVED** [review]
